radix-engine/src/engine/kernel.rs

```diff
@@ -178,32 +180,14 @@ where
                 ))
             })?;
 
-        if let SubstateId::ComponentInfo(address) = substate_id {
-            node_pointer
-                .acquire_lock(SubstateId::ComponentState(*address), false, false, track)
-                .map_err(RuntimeError::KernelError)?;
-            node_pointer
-                .acquire_lock(SubstateId::ComponentInfo(*address), false, false, track)
-                .map_err(RuntimeError::KernelError)?;
-        }
-
         // Read current value
-        let current_value = {
+        let substate = {
             let mut node_ref = node_pointer.to_ref_mut(call_frames, track);
-            node_ref.read_scrypto_value(&substate_id)?
+            node_ref.read_substate(&substate_id)?
         };
+        let contained_value = extract_value_from_substate(substate_id, &substate);
```

Comment:
Seems like `extract_value_from_substate` should be managed by the caller of this method rather than in `read_substate_internal`

Reply:
Updated.